File: python_compiler/env.py
from settings import *

class Env:
    """Definitions are referenced with zero-based indices. Each class of definitions has its own index space, as distinguished
    by the following classes"""

    def __init__(self):
        # Unnamed indices are associated with the empty (\epsilon) entries in these lists
        self.types = {}
        self.funcs = {}
        self.tables = {}
        self.mems = {}
        self.globals = {}
        self.elems = {}
        self.data = {}
        self.typedefs = {}

    def _getContext(self, space):
        if space == "types":
            return self.types
        elif space == "funcs":
            return self.funcs
        elif space == "tables":
            return self.tables
        elif space == "mems":
            return self.mems
        elif space == "globals":
            return self.globals
        elif space == "data":
            return self.data
        elif space == "elems":
            return self.elems
        elif space == "typedefs":
            return self.typedefs

    def getIdentifierIndex(self, space, identifier):
        """Returns the index corresponding to the identifier"""
        if isinstance(identifier, str):
            for index,iteratedIdentifier in self._getContext(space).items():
                if iteratedIdentifier == identifier: return index
        elif isinstance(identifier, int):
            return identifier
    def addIdentifierIndex(self, space, identifier):
        """Add index corresponding to the identifier"""
        index = len(self._getContext(space))
        self._getContext(space)[index] = identifier
        return index

class Frame:
    def __init__(self):
        # Unnamed indices are associated with the empty (\epsilon) entries in these lists
        self.locals = {}
        self.labels = {}
        self.returns = []
        self.args = {}

    def returnArity(self):
        return len(self.returns)
    def setLocal(self, value):
        index = len(self.locals)
        self.locals[index] = value
    def setLocalValue(self, value, index):
        self.locals[index]['value'] = value
    def getLocal(self, index):
        return self.locals[index]['value']
    def addArg(self, arg):
        index = len(self.args)
        self.args[index] = arg
    def addReturn(self, valtype):
        self.returns.append(valtype)
    def argArity(self):
        return len(self.args)
    

class Label:
    def __init__(self, instructions, argumentArity, index, type):
        self.instructions = instructions
        self.argumentArity = argumentArity
        self.index = index
        self.type = type # BLOCK, LOOP
        self.continuation = []
        # If it is of type loop, then the continuation is all instructions leading up to the branch instruction
        if self.type == LOOP:
            # Loop through the instructions and while there is no branch instruction, add to the continuation
            for i in range(len(self.instructions)):
                self.continuation.append(self.instructions[i])
                if self.instructions[i]['type'] == BR or self.instructions[i]['type'] == BR_IF:
                    break
    

File: go_compiler/parser/env.go
package parser

import "main/scanner"

// Create the methods for our environment method
type EnvInterface interface {
	getContext(context string) map[int]*string 
    getIdentifierContext(identifierType string, identifier interface{}) interface{}
    addIdentifierContext(identifierType string, identifier *string) int
}

// Struct to hold the identifier context for our module. dictionary represents and index space. When we would like to know
// What an id is referencing, we look for it inside the environment
type Env struct {
	Types map[int]*string
    Funcs map[int]*string
    Tables map[int]*string
    Mems map[int]*string
    Globals map[int]*string
    Elems map[int]*string
    Data map[int]*string
    Typedefs map[int]*string
}

// Method to get the context pertaining to a string input. They are now able to read and write
// to the corresponding context.
func (e Env) getContext(identifierType int) *map[int]*string  {
    switch identifierType {
    case scanner.TYPE:
        return &e.Types
    case scanner.FUNC:
        return &e.Funcs
    case scanner.TABLE:
        return &e.Tables
    case scanner.MEMORY:
        return &e.Mems
    case scanner.GLOBAL:
        return &e.Globals
    }
    return nil
}

// Method to get the corresponding identifier index for a certain context space
func (e Env) GetIdentifierContextIndex(identifierSpace int, identifier interface{}) *int {
    identifierTypeMap := *e.getContext(identifierSpace)
    switch identifier.(type) {
    case int:
        index, _ := identifier.(int)
        return &index
    case string:
        // Iterate through the map
        for index, indexedIdentifier := range identifierTypeMap {
            if *indexedIdentifier == identifier.(string) {
                return &index
            }
        }
    }
    return nil
}

// Method to add an identifier to an identifier context 
func (e *Env) AddIdentifierContext(identifierSpace int, identifier *string) int {
    identifierTypeMap := *e.getContext(identifierSpace)
    index := len(identifierTypeMap)
    identifierTypeMap[index] = identifier
    return index
}

// Struct to hold our environment for functions
type Frame struct {
    Locals []LocalNode
    labels []int
    returns []ReturnNode
    args []LocalNode

}

// Holds the methods that will interact with our frame
type FrameInterface interface {
    ReturnArity() int
    SetLocal(value int)
    SetLocalValue(value int, index int)
    GetLocal(index int)
    AddArg(arg LocalNode)
    AddReturn(valType int)
    ArgArity() int
    LocalArity() int
}

// Return the number of returns for the frame
func (f *Frame) ReturnArity()int {
    return len(f.returns)
}

// Add a new local node to our list of locals
func (f *Frame) SetLocal(value LocalNode) {
    f.Locals = append(f.Locals, value)
}

/* Add a new value to a certain local node
    value - value to set it to
    index - position in locals
*/
func (f *Frame) SetLocalValue(value int, index int) {
    f.Locals[index].Value = value
}

// Get the value associated with a certain local
func (f *Frame) GetLocal(index int) int {
    return f.Locals[index].Value
}

// Add an argument to the list of arguments
func (f *Frame) AddArgument(arg LocalNode) {
    f.args = append(f.args, arg)
}

// Add a return value to the list of returns
func (f *Frame) AddReturn(valtype int) {
    f.returns = append(f.returns, ReturnNode{TokenType:valtype})
}

// Get the arity of the arguments
func (f *Frame) ArgArity() int {
    return len(f.args)
}

// Label will hold information regarding our block instruction
type Label struct {
    Instructions []interface{} 
    ArgumentArity int
    Index int
    LabelType int
}

